### PR TITLE
Stage and job

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

File: engine/api/pipeline_job.go [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+
+	pipelineData, err := pipeline.LoadPipeline(db, key, pipName, false)
+	if err != nil {
+		log.Warning("updateJobHandler>Cannot load pipeline %s: %s\n", pipName, err)
+		WriteError(w, r, err)
+		return
+	}
+
+	if err := pipeline.LoadPipelineStage(db, pipelineData); err != nil {
+		log.Warning("deletepipelineActionHandler>Cannot load stages: %s\n", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> updateJobHandler

File: engine/api/pipeline_job.go [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+			for _, j := range s.Actions {
+				if j.PipelineActionID == jobID {
+					found = true
+				}
+			}
+		}
+	}
+
+	if !found {
+		log.Warning("deletepipelineActionHandler>Job not found\n")
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> updateJobHandler

File: engine/api/pipeline_job.go [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+	// Get pipeline and action name in URL
+	vars := mux.Vars(r)
+	key := vars["key"]
+	pipName := vars["permPipelineKey"]
+	stageIDString := vars["stageID"]
+	jobIDString := vars["jobID"]
+
+	jobID, err := strconv.ParseInt(jobIDString, 10, 64)
+	if err != nil {
+		log.Warning("deletepipelineActionHandler>ID is not a int: %s\n", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> deleteJobHandler

File: engine/api/pipeline_job.go [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+		WriteError(w, r, sdk.ErrInvalidID)
+		return
+	}
+
+	stageID, err := strconv.ParseInt(stageIDString, 10, 64)
+	if err != nil {
+		log.Warning("deletepipelineActionHandler>ID is not a int: %s\n", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> deleteJobHandler

File: engine/api/pipeline_job.go [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+		WriteError(w, r, sdk.ErrInvalidID)
+		return
+	}
+
+	pipelineData, err := pipeline.LoadPipeline(db, key, pipName, false)
+	if err != nil {
+		log.Warning("deletepipelineActionHandler>Cannot load pipeline %s: %s\n", pipName, err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> deleteJobHandler

File: engine/api/pipeline_job.go [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+		WriteError(w, r, err)
+		return
+	}
+
+	if err := pipeline.LoadPipelineStage(db, pipelineData); err != nil {
+		log.Warning("deletepipelineActionHandler>Cannot load stages: %s\n", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> deleteJobHandler

File: engine/api/pipeline_job.go [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+		return
+	}
+
+	// check if job is in the current pipeline
+	found := false
+	for _, s := range pipelineData.Stages {
+		if s.ID == stageID {
+			for _, j := range s.Actions {
+				if j.PipelineActionID == jobID {
+					found = true
```
Comment (proullon) [inline engine/api/pipeline_job.go]: break

File: engine/api/pipeline_job.go [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
```
Comment (proullon) [inline engine/api/pipeline_job.go]: break

File: engine/api/pipeline_job.go [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+
+
+	// Select all pipeline build where given pipelineAction has been run
+	query := `SELECT pipeline_build.id FROM pipeline_build
+						JOIN action_build ON action_build.pipeline_build_id = pipeline_build.id
+						WHERE action_build.pipeline_action_id = $1`
+	var ids []int64
+	rows, err := db.Query(query, jobID)
+	if err != nil {
+		log.Warning("deletePipelineActionHandler> cannot retrieves pipeline build: %s\n", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> deleteJobHandler

File: engine/api/pipeline_job.go [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+		WriteError(w, r, err)
+		return
+	}
+	defer tx.Rollback()
+
+	// For each pipeline build, archive it to get out of relationnal
+	for _, id := range ids {
+		err = archivist.ArchiveBuild(tx, id)
+		if err != nil {
+			log.Warning("deletePipelineActionHandler> cannot archive pipeline build: %s\n", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> deleteJobHandler

File: engine/api/pipeline_job.go [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+			WriteError(w, r, err)
+			return
+		}
+	}
+
+	err = pipeline.DeleteJob(tx, jobID, c.User.ID)
+	if err != nil {
+		log.Warning("deletePipelineActionHandler> Cannot delete pipeline action: %s", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> deleteJobHandler

File: engine/api/pipeline_job.go [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+		WriteError(w, r, err)
+		return
+	}
+
+	err = pipeline.UpdatePipelineLastModified(tx, pipelineData)
+	if err != nil {
+		log.Warning("deletePipelineActionHandler> Cannot update pipeline last_modified: %s", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> deleteJobHandler

File: engine/api/pipeline_job.go [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+		WriteError(w, r, err)
+		return
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		log.Warning("deletePipelineActionHandler> Cannot commit transaction: %s", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> deleteJobHandler

File: engine/api/pipeline_job.go [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+
+	pip, err := pipeline.LoadPipeline(db, projectKey, pipelineName, false)
+	if err != nil {
+		log.Warning("addJobToStageHandler> Cannot load pipeline %s for project %s: %s\n", pipelineName, projectKey, err)
+		WriteError(w, r, sdk.ErrPipelineNotFound)
+		return
+	}
+
+	if err := pipeline.LoadPipelineStage(db, pip); err != nil {
+		log.Warning("deletepipelineActionHandler>Cannot load stages: %s\n", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> addJobToStageHandler

File: engine/api/pipeline_job.go [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+	// check if stage is in the current pipeline
+	found := false
+	for _, s := range pip.Stages {
+		if s.ID == stageID {
+			found = true
+		}
+	}
+
+	if !found {
+		log.Warning("deletepipelineActionHandler>Stage not found\n")
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> addJobToStageHandler

File: engine/api/pipeline_job.go [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+		log.Warning("deletepipelineActionHandler>Cannot load stages: %s\n", err)
+		WriteError(w, r, err)
+		return
+	}
+
+	// check if stage is in the current pipeline
+	found := false
+	for _, s := range pip.Stages {
+		if s.ID == stageID {
+			found = true
```
Comment (proullon) [inline engine/api/pipeline_job.go]: break

File: engine/api/pipeline_job.go [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+	jobID, err := strconv.ParseInt(jobIDString, 10, 64)
+	if err != nil {
+		log.Warning("updateJobHandler>ID is not a int: %s\n", err)
+		WriteError(w, r, sdk.ErrInvalidID)
+		return
+	}
+
+	stageID, err := strconv.ParseInt(stageIDString, 10, 64)
+	if err != nil {
+		log.Warning("deletepipelineActionHandler>ID is not a int: %s\n", err)
```
Comment (proullon) [inline engine/api/pipeline_job.go]: deletepipelineActionHandler -> updateJobHandler

**Comment by sguiheux** [comment]
All done + one line + delete job

File: engine/api/main.go [hunk 18]
```diff
@@ -315,10 +315,14 @@ func (router *Router) init() {
...
 	router.Handle("/project/{key}/pipeline/{permPipelineKey}", GET(getPipelineHandler), PUT(updatePipelineHandler), DELETE(deletePipeline))
-	router.Handle("/project/{key}/pipeline/{permPipelineKey}/action/{pipelineActionID}", PUT(updatePipelineActionHandler), DELETE(deletePipelineActionHandler))
 	router.Handle("/project/{key}/pipeline/{permPipelineKey}/stage", POST(addStageHandler))
 	router.Handle("/project/{key}/pipeline/{permPipelineKey}/stage/move", POST(moveStageHandler))
 	router.Handle("/project/{key}/pipeline/{permPipelineKey}/stage/{stageID}", GET(getStageHandler), PUT(updateStageHandler), DELETE(deleteStageHandler))
+	router.Handle("/project/{key}/pipeline/{permPipelineKey}/stage/{stageID}/job", POST(addJobToStageHandler))
+	router.Handle("/project/{key}/pipeline/{permPipelineKey}/stage/{stageID}/job/{jobID}", PUT(updateJobHandler), DELETE(deleteJobHandler))
+
+	// DEPRECATED
+	router.Handle("/project/{key}/pipeline/{permPipelineKey}/action/{pipelineActionID}", PUT(updatePipelineActionHandler), DELETE(deletePipelineActionHandler))
```
Comment (fsamin) [inline engine/api/main.go]: Please  specify which handler is deprecated

File: engine/api/action/action.go [hunk 19]
```diff
@@ -275,14 +275,14 @@ func UpdateActionDB(tx *sql.Tx, a *sdk.Action, userID int64) error {
...
+		err = InsertActionRequirement(db, a.ID, a.Requirements[i])
 		if err != nil {
 			return err
 		}
 	}
 
 	query := `UPDATE action SET name=$1,description=$2, type=$3, enabled=$4 WHERE id=$5`
-	_, err = tx.Exec(query, a.Name, a.Description, string(a.Type), a.Enabled, a.ID)
+	_, err = db.Exec(query, a.Name, a.Description, string(a.Type), a.Enabled, a.ID)
 	if err != nil {
```
Comment (fsamin) [inline engine/api/action/action.go]: one line

File: engine/api/action/action.go [hunk 20]
```diff
@@ -275,14 +275,14 @@ func UpdateActionDB(tx *sql.Tx, a *sdk.Action, userID int64) error {
 		}
 	}
 	for i := range a.Requirements {
-		err = InsertActionRequirement(tx, a.ID, a.Requirements[i])
+		err = InsertActionRequirement(db, a.ID, a.Requirements[i])
```
Comment (fsamin) [inline engine/api/action/action.go]: one line

File: engine/api/action/action.go [hunk 21]
```diff
@@ -208,53 +208,53 @@ func loadAction(db database.Querier, s database.Scanner) (*sdk.Action, error) {
...
-		err = InsertActionParameter(tx, a.ID, a.Parameters[i])
+		err = InsertActionParameter(db, a.ID, a.Parameters[i])
 		if err != nil {
 			log.Warning("UpdateAction> InsertActionParameter for %s failed: %s\n", a.Parameters[i].Name, err)
 			return err
 		}
 	}
 
-	err = DeleteActionRequirements(tx, a.ID)
+	err = DeleteActionRequirements(db, a.ID)
```
Comment (fsamin) [inline engine/api/action/action.go]: one line

File: engine/api/action/action.go [hunk 22]
```diff
@@ -208,53 +208,53 @@ func loadAction(db database.Querier, s database.Scanner) (*sdk.Action, error) {
...
 	}
 
-	err = DeleteActionParameters(tx, a.ID)
+	err = DeleteActionParameters(db, a.ID)
 	if err != nil {
 		return err
 	}
 	for i := range a.Parameters {
-		err = InsertActionParameter(tx, a.ID, a.Parameters[i])
+		err = InsertActionParameter(db, a.ID, a.Parameters[i])
```
Comment (fsamin) [inline engine/api/action/action.go]: one line

File: engine/api/action/action.go [hunk 23]
```diff
@@ -208,53 +208,53 @@ func loadAction(db database.Querier, s database.Scanner) (*sdk.Action, error) {
...
 		}
 
-		if err = insertActionChild(tx, a.ID, a.Actions[i], i+1); err != nil {
+		if err = insertActionChild(db, a.ID, a.Actions[i], i+1); err != nil {
 			return err
 		}
 	}
 
-	err = DeleteActionParameters(tx, a.ID)
+	err = DeleteActionParameters(db, a.ID)
```
Comment (fsamin) [inline engine/api/action/action.go]: one line

File: engine/api/action/action.go [hunk 24]
```diff
@@ -208,53 +208,53 @@ func loadAction(db database.Querier, s database.Scanner) (*sdk.Action, error) {
...
 	}
 
-	err = insertAudit(tx, a.ID, userID, "Action update")
+	err = insertAudit(db, a.ID, userID, "Action update")
 	if err != nil {
 		return err
 	}
 
-	err = deleteActionChildren(tx, a.ID)
+	err = deleteActionChildren(db, a.ID)
```
Comment (fsamin) [inline engine/api/action/action.go]: one line

File: engine/api/action/action.go [hunk 25]
```diff
@@ -208,53 +208,53 @@ func loadAction(db database.Querier, s database.Scanner) (*sdk.Action, error) {
...
+	ok, err := isTreeLoopFree(db, a, nil)
 	if err != nil {
 		return err
 	}
 	if !ok {
 		return sdk.ErrActionLoop
 	}
 
-	err = insertAudit(tx, a.ID, userID, "Action update")
+	err = insertAudit(db, a.ID, userID, "Action update")
```
Comment (fsamin) [inline engine/api/action/action.go]: one line

File: engine/api/pipeline/pipeline_action.go [hunk 26]
```diff
@@ -128,7 +131,70 @@ func InsertPipelineAction(db database.QueryExecuter, projectKey, pipelineName st
 		return 0, err
 	}
 
-	return a.PipelineActionID, UpdatePipelineLastModified(db, p.ID)
+	return a.PipelineActionID, UpdatePipelineLastModified(db, p)
+}
+
+// InsertJob  Insert a new Job ( pipeline_action + joinedAction )
+func InsertJob(db database.QueryExecuter, job *sdk.Job, stageID int64, pip *sdk.Pipeline) error {
```
Comment (fsamin) [inline engine/api/pipeline/pipeline_action.go]: Where is the unit test ?

File: engine/api/pipeline/pipeline_action.go [hunk 27]
```diff
@@ -128,7 +131,70 @@ func InsertPipelineAction(db database.QueryExecuter, projectKey, pipelineName st
...
+	// Create pipeline action
+	query := `INSERT INTO pipeline_action (pipeline_stage_id, action_id, enabled) VALUES ($1, $2, $3) RETURNING id`
+	if err := db.QueryRow(query, job.PipelineStageID, job.Action.ID, job.Enabled).Scan(&job.PipelineActionID); err != nil {
+		return err
+	}
+	return nil
+}
+
+// UpdateJob  updates the job by actionData.PipelineActionID and actionData.ID
+func UpdateJob(db database.QueryExecuter, job *sdk.Job, userID int64) error {
```
Comment (fsamin) [inline engine/api/pipeline/pipeline_action.go]: Where is the unit test ?

File: engine/api/pipeline_job.go [hunk 28]
```diff
@@ -0,0 +1,378 @@
...
+	"github.com/ovh/cds/engine/api/context"
+	"github.com/ovh/cds/engine/api/pipeline"
+	"github.com/ovh/cds/engine/api/project"
+	"github.com/ovh/cds/engine/api/sanity"
+	"github.com/ovh/cds/engine/log"
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/engine/api/archivist"
+)
+
+func addJobToStageHandler(w http.ResponseWriter, r *http.Request, db *sql.DB, c *context.Context) {
```
Comment (fsamin) [inline engine/api/pipeline_job.go]: Where is the unit test ?

File: engine/api/pipeline_job.go [hunk 29]
```diff
@@ -0,0 +1,378 @@
...
+	for _, s := range pip.Stages {
+		if s.ID == stageID {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		log.Warning("addJobToStageHandler>Stage not found\n")
+		WriteError(w, r, sdk.ErrWrongRequest)
```
Comment (fsamin) [inline engine/api/pipeline_job.go]: 404 error ?

File: engine/api/pipeline_job.go [hunk 30]
```diff
@@ -0,0 +1,378 @@
...
+		return
+	}
+
+	if err := pipeline.UpdatePipelineLastModified(tx, pip); err != nil {
+		log.Warning("addJobToStageHandler> Cannot update pipeline last modified date: %s\n", err)
+		WriteError(w, r, err)
+		return
+	}
+
+	//warnings, err := sanity.CheckActionRequirements(tx, proj.Key, pip.Name, a.ID)
```
Comment (fsamin) [inline engine/api/pipeline_job.go]: do not compute warning because the job is empty

File: engine/api/pipeline_job.go [hunk 31]
```diff
@@ -0,0 +1,378 @@
...
+	if err := pipeline.LoadPipelineStage(db, pipelineData); err != nil {
+		log.Warning("updateJobHandler> Cannot load stages: %s\n", err)
+		WriteError(w, r, err)
+		return
+	}
+
+	WriteJSON(w, r, pipelineData, http.StatusOK)
+}
+
+func deleteJobHandler(w http.ResponseWriter, r *http.Request, db *sql.DB, c *context.Context) {
```
Comment (fsamin) [inline engine/api/pipeline_job.go]: unit tests

File: engine/api/pipeline_job.go [hunk 32]
```diff
@@ -0,0 +1,378 @@
...
+					found = true
+					break
+				}
+			}
+		}
+	}
+
+	if !found {
+		log.Warning("deleteJobHandler>Job not found\n")
+		WriteError(w, r, sdk.ErrWrongRequest)
```
Comment (fsamin) [inline engine/api/pipeline_job.go]: 404 ?

File: engine/api/pipeline_job.go [hunk 33]
```diff
@@ -0,0 +1,378 @@
...
+		return
+	}
+
+
+
+	// Select all pipeline build where given pipelineAction has been run
+	query := `SELECT pipeline_build.id FROM pipeline_build
```
Comment (fsamin) [inline engine/api/pipeline_job.go]: i think we shouldn't see sql in handlers

File: engine/api/pipeline_job.go [hunk 34]
```diff
@@ -0,0 +1,378 @@
...
+	rows, err := db.Query(query, jobID)
+	if err != nil {
+		log.Warning("deleteJobHandler> cannot retrieves pipeline build: %s\n", err)
+		WriteError(w, r, err)
+		return
+	}
+
+	for rows.Next() {
+		var id int64
+		err = rows.Scan(&id)
```
Comment (fsamin) [inline engine/api/pipeline_job.go]: one line

File: engine/api/pipeline_job.go [hunk 35]
```diff
@@ -0,0 +1,378 @@
...
+						JOIN action_build ON action_build.pipeline_build_id = pipeline_build.id
+						WHERE action_build.pipeline_action_id = $1`
+	var ids []int64
+	rows, err := db.Query(query, jobID)
+	if err != nil {
+		log.Warning("deleteJobHandler> cannot retrieves pipeline build: %s\n", err)
+		WriteError(w, r, err)
+		return
+	}
+
```
Comment (fsamin) [inline engine/api/pipeline_job.go]: defer row.Close() and remove row.Close() below

File: engine/api/pipeline_job.go [hunk 36]
```diff
@@ -0,0 +1,378 @@
...
+	if err != nil {
+		log.Warning("deleteJobHandler> Cannot begin transaction: %s\n", err)
+		WriteError(w, r, err)
+		return
+	}
+	defer tx.Rollback()
+
+	// For each pipeline build, archive it to get out of relationnal
+	for _, id := range ids {
+		err = archivist.ArchiveBuild(tx, id)
```
Comment (fsamin) [inline engine/api/pipeline_job.go]: one line

File: engine/api/pipeline_parameter.go [hunk 37]
```diff
@@ -67,7 +67,7 @@ func deleteParameterFromPipelineHandler(w http.ResponseWriter, r *http.Request,
 		return
 	}
 
-	err = pipeline.UpdatePipelineLastModified(tx, p.ID)
+	err = pipeline.UpdatePipelineLastModified(tx, p)
```
Comment (fsamin) [inline engine/api/pipeline_parameter.go]: one line

File: engine/api/pipeline_parameter.go [hunk 38]
```diff
@@ -198,7 +198,7 @@ func updateParametersInPipelineHandler(w http.ResponseWriter, r *http.Request, d
 		return
 	}
 
-	err = pipeline.UpdatePipelineLastModified(tx, pip.ID)
+	err = pipeline.UpdatePipelineLastModified(tx, pip)
```
Comment (fsamin) [inline engine/api/pipeline_parameter.go]: one line

File: engine/api/pipeline_parameter.go [hunk 39]
```diff
@@ -270,7 +270,7 @@ func updateParameterInPipelineHandler(w http.ResponseWriter, r *http.Request, db
 		}
 	}
 
-	err = pipeline.UpdatePipelineLastModified(tx, p.ID)
+	err = pipeline.UpdatePipelineLastModified(tx, p)
 	if err != nil {
```
Comment (fsamin) [inline engine/api/pipeline_parameter.go]: one line

File: engine/api/pipeline_parameter.go [hunk 40]
```diff
@@ -346,7 +346,7 @@ func addParameterInPipelineHandler(w http.ResponseWriter, r *http.Request, db *s
 		}
 	}
 
-	err = pipeline.UpdatePipelineLastModified(tx, p.ID)
+	err = pipeline.UpdatePipelineLastModified(tx, p)
```
Comment (fsamin) [inline engine/api/pipeline_parameter.go]: one line

File: engine/api/stage.go [hunk 41]
```diff
@@ -300,7 +329,7 @@ func deleteStageHandler(w http.ResponseWriter, r *http.Request, db *sql.DB, c *c
 		return
 	}
 
-	err = pipeline.UpdatePipelineLastModified(tx, pipelineData.ID)
+	err = pipeline.UpdatePipelineLastModified(tx, pipelineData)
 	if err != nil {
```
Comment (fsamin) [inline engine/api/stage.go]: one line

File: engine/api/main.go (hunk 18 above)
Comment (sguiheux) [inline engine/api/main.go]: There is space between and after, so for me it's ok.

/audit will be rewrite soon

**Pull request merged**